File: competitive_programming/leetCode/problem_list/Array/easy/169_majorityElement.go
package easy

/**
Problem: https://leetcode.com/problems/majority-element/
*/

/*
Time: O(n)
Space: O(n)
*/

func MajorityElement_usingHashMap(nums []int) int {
	// Create a hash map to store the frequency of each element
	frequencyMap := make(map[int]int)

	// Iterate through the array to count the frequency of each element
	for _, num := range nums {
		frequencyMap[num]++

		// Check if the current element is the majority element
		if frequencyMap[num] > len(nums)/2 {
			return num
		}
	}

	// The majority element always exists, so this line should not be reached.
	return -1
}

func MajorityElement(nums []int) int {
	majority := nums[0]
	count := 1

	for i := 1; i < len(nums); i++ {
		if nums[i] == majority {
			count++
		} else {
			count--
			if count == 0 {
				majority = nums[i]
				count = 1
			}
		}
	}
	return majority
}

File: ruby/easy/169_majority_element.rb
def majority_element(nums)
    candidate = nil
    count = 0
  
    # Phase 1: Find a candidate
    nums.each do |num|
      if count == 0
        candidate = num
      end
      count += (num == candidate ? 1 : -1)
    end
  
    # Since the problem guarantees that a majority element always exists,
    # we can directly return the candidate.
    candidate
end
  